File: src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatEncryption.java
package me.zeroeightsix.kami.module.modules.chat;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.KamiMod;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.client.CPacketChatMessage;
import net.minecraft.network.play.server.SPacketChat;
import net.minecraft.util.ChatAllowedCharacters;
import net.minecraft.util.text.TextComponentString;

import java.nio.CharBuffer;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static me.zeroeightsix.kami.command.Command.getCommandPrefix;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendErrorMessage;

/**
 * Created by 086 on 9/04/2018.
 * Updated by dominikaaaa on 20/04/20
 */
@Module.Info(
        name = "ChatEncryption",
        description = "Encrypts and decrypts chat messages",
        category = Module.Category.CHAT,
        showOnArray = Module.ShowOnArray.OFF
)
public class ChatEncryption extends Module {

    private Setting<Boolean> self = register(Settings.b("Decrypt own", true));
    private Setting<EncryptionMode> mode = register(Settings.e("Mode", EncryptionMode.SHUFFLE));
    private Setting<Integer> key = register(Settings.i("Key", 6));
    private Setting<Boolean> delim = register(Settings.b("Delimiter", true));
    public static Setting<String> delimiterValue = Settings.s("delimiterV", "unchanged");

    private final Pattern CHAT_PATTERN = Pattern.compile("<.*?> ");

    private static final char[] ORIGIN_CHARS = new char[]{'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '-', '_', '/', ';', '=', '?', '+', '\u00B5', '\u00A3', '*', '^', '\u00F9', '$', '!', '{', '}', '\'', '"', '|', '&'};

    @EventHandler
    private Listener<PacketEvent.Send> sendListener = new Listener<>(event -> {
        if (event.getPacket() instanceof CPacketChatMessage) {
            String s = ((CPacketChatMessage) event.getPacket()).getMessage();
            if (delim.getValue()) {
                if (getDelimiter() == null || !s.startsWith(getDelimiter())) return;
                s = s.substring(1);
            }
            StringBuilder builder = new StringBuilder();
            switch (mode.getValue()) {
                case SHUFFLE:
                    builder.append(shuffle(key.getValue(), s));
                    builder.append("\uD83D\uDE4D");
                    break;
                case SHIFT:
                    s.chars().forEachOrdered(value -> builder.append((char) (value + (ChatAllowedCharacters.isAllowedCharacter((char) (value + key.getValue())) ? key.getValue() : 0))));
                    builder.append("\uD83D\uDE48");
                    break;
            }
            s = builder.toString();
            if (s.length() > 256) {
                sendChatMessage("Encrypted message length was too long, couldn't send!");
                event.cancel();
                return;
            }
            ((CPacketChatMessage) event.getPacket()).message = s;
        }
    });

    @EventHandler
    private Listener<PacketEvent.Receive> receiveListener = new Listener<>(event -> {
        if (event.getPacket() instanceof SPacketChat) {
            String s = ((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText();

            if (!self.getValue() && isOwn(s)) return;

            Matcher matcher = CHAT_PATTERN.matcher(s);
            String username = "unnamed";
            if (matcher.find()) {
                username = matcher.group();
                username = username.substring(1, username.length() - 2);
                s = matcher.replaceFirst("");
            }

            StringBuilder builder = new StringBuilder();
            String substring = s.substring(0, s.length() - 2);
            switch (mode.getValue()) {
                case SHUFFLE:
                    if (!s.endsWith("\uD83D\uDE4D")) return;
                    s = substring;
                    builder.append(unshuffle(key.getValue(), s));
                    break;
                case SHIFT:
                    if (!s.endsWith("\uD83D\uDE48")) return;
                    s = substring;
                    s.chars().forEachOrdered(value -> builder.append((char) (value + (ChatAllowedCharacters.isAllowedCharacter((char) value) ? -key.getValue() : 0))));
                    break;
            }

            ((SPacketChat) event.getPacket()).chatComponent = new TextComponentString("<" + username + "> " + KamiMod.colour + "lDECRYPTED" + KamiMod.colour + "r: " + builder.toString());
        }
    });

    private Map<Character, Character> generateShuffleMap(int seed) {
        Random r = new Random(seed);
        List<Character> characters = CharBuffer.wrap(ORIGIN_CHARS).chars().mapToObj(value -> (char) value).collect(Collectors.toList());
        List<Character> counter = new ArrayList<>(characters);
        Collections.shuffle(counter, r);

        Map<Character, Character> map = new LinkedHashMap<>();  // ordered
        for (int i = 0; i < characters.size(); i++) {
            map.put(characters.get(i), counter.get(i));
        }

        return map;
    }

    private String shuffle(int seed, String input) {
        Map<Character, Character> s = generateShuffleMap(seed);
        StringBuilder builder = new StringBuilder();
        swapCharacters(input, s, builder);
        return builder.toString();
    }

    private String unshuffle(int seed, String input) {
        Map<Character, Character> s = generateShuffleMap(seed);
        StringBuilder builder = new StringBuilder();
        swapCharacters(input, reverseMap(s), builder);
        return builder.toString();
    }

    private void swapCharacters(String input, Map<Character, Character> s, StringBuilder builder) {
        CharBuffer.wrap(input.toCharArray()).chars().forEachOrdered(value -> {
            char c = (char) value;
            if (s.containsKey(c)) builder.append(s.get(c));
            else
                builder.append(c);
        });
    }

    private static <K, V> Map<V, K> reverseMap(Map<K, V> map) {
        return map.entrySet().stream().collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
    }

    private enum EncryptionMode {
        SHUFFLE, SHIFT
    }

    private boolean isOwn(String message) {
        return Pattern.compile("^<" + mc.player.getName() + "> ", Pattern.CASE_INSENSITIVE).matcher(message).find();
    }

    private String getDelimiter() {
        if (delimiterValue.getValue().equalsIgnoreCase("unchanged")) {
            sendErrorMessage(getChatName() + " Please change the delimiter with &7" + getCommandPrefix() + "chatencryption&f, disabling");
            disable();
            return null;
        }
        return delimiterValue.getValue();
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatEncryption.kt
package me.zeroeightsix.kami.module.modules.chat

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.manager.managers.MessageManager
import me.zeroeightsix.kami.manager.managers.MessageManager.newMessageModifier
import me.zeroeightsix.kami.mixin.extension.textComponent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.text.MessageDetection
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.text.format
import me.zeroeightsix.kami.util.text.formatValue
import me.zeroeightsix.kami.util.threads.defaultScope
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.network.play.server.SPacketChat
import net.minecraft.util.text.TextFormatting
import org.kamiblue.commons.utils.SystemUtils
import java.util.*
import kotlin.collections.HashMap

// TODO: Add proper RSA encryption
@Module.Info(
    name = "ChatEncryption",
    description = "Encrypts and decrypts chat messages",
    category = Module.Category.CHAT,
    modulePriority = -69420
)
object ChatEncryption : Module() {
    private val commands = setting("Commands", false)
    private val self = setting("DecryptOwn", true)
    private val keySetting = setting("KeySetting", "DefaultKey")
    val delimiter = setting("Delimiter", "%", consumer = { prev: String, value: String ->
        if (value.length == 1 && !chars.contains(value.first())) value else prev
    })

    private const val personFrowning = "🙍"

    private val chars = charArrayOf(
        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
        'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
        ',', '.', '!', '?', ' '
    )

    private val charsMap: Map<Char, Int> = HashMap<Char, Int>(67).apply {
        for ((index, char) in chars.withIndex()) {
            put(char, index)
        }
    }

    private val vigenereSquare = Array(chars.size) { index1 ->
        CharArray(chars.size) { index2 ->
            chars[(index1 + index2) % chars.size]
        }
    }

    private val reversedSquare = Array(chars.size) { index1 ->
        CharArray(chars.size) { index2 ->
            chars[Math.floorMod((index1 - index2), chars.size)]
        }
    }

    private val modifier = newMessageModifier(
        filter = {
            (commands.value || MessageDetection.Command.ANY_EXCEPT_DELIMITER detectNot it.packet.message)
        },
        modifier = {
            it.encrypt() ?: it.packet.message
        }
    )

    private var previousMessage = ""

    override fun onEnable() {
        getOrGenKey()
        modifier.enable()
    }

    override fun onDisable() {
        modifier.disable()
    }

    private fun getOrGenKey(): String {
        var key = keySetting.value

        if (key == "DefaultKey") {
            key = randomChars()
            keySetting.value = key

            MessageSendHelper.sendChatMessage("$chatName Your encryption key was set to ${formatValue(key)}, and copied to your clipboard.")

            defaultScope.launch(Dispatchers.IO) {
                SystemUtils.copyToClipboard(key)
            }
        }

        return key
    }

    private fun randomChars() = StringBuilder().run {
        for (i in 1..32) {
            append(chars.random())
        }
        toString()
    }

    init {
        safeListener<PacketEvent.Receive> {
            if (it.packet !is SPacketChat) return@safeListener

            val fullMessage: CharSequence = it.packet.textComponent.unformattedText
            if (!fullMessage.contains(personFrowning)) return@safeListener

            val playerName = MessageDetection.Message.ANY.playerName(fullMessage) ?: "Unknown User"
            if (!self.value && playerName == player.name) return@safeListener

            val message = MessageDetection.Message.ANY.removedOrNull(fullMessage) ?: return@safeListener
            val splitString = message.split(personFrowning)

            val final = StringBuilder().run {
                for ((index, string) in splitString.withIndex()) {
                    if (index % 2 == 0) {
                        append(string)
                    } else {
                        decodeVigenere(string)
                    }
                }
                toString()
            }

            if (final == message) return@safeListener // prevent further possible issues

            MessageSendHelper.sendRawChatMessage("<$playerName> ${TextFormatting.BOLD format "DECRYPTED"}: $final")
        }
    }

    private fun MessageManager.QueuedMessage.encrypt(): String? {
        val message = packet.message

        // fix existing issue - for some reason this is run twice sometimes (single player only)
        if (message == previousMessage) {
            previousMessage = ""
            return null
        }

        val splitString = message.split(delimiter.value)

        val encrypted = StringBuilder().run {
            for ((index, string) in splitString.withIndex()) {
                if (index % 2 == 0) {
                    append(string)
                } else {
                    append(personFrowning)
                    encodeVigenere(string)
                    append(personFrowning)
                }
            }
            toString()
        }

        if (encrypted.isBlank()) return null

        if (encrypted.length > 256) { // this shouldn't happen unless your message was already around 254 chars?
            MessageSendHelper.sendChatMessage("Encrypted message length was too long, couldn't send!")
            return null
        }

        previousMessage = encrypted
        return encrypted
    }

    private fun StringBuilder.encodeVigenere(string: String) {
        val key = getOrGenKey()

        for ((index, char) in string.withIndex()) {
            val indexOfKey = charsMap[key[index % key.length]] ?: continue
            val indexOfChar = charsMap[char] ?: continue
            val encodedChar = vigenereSquare[indexOfKey][indexOfChar]

            append(encodedChar)
        }
    }

    private fun StringBuilder.decodeVigenere(string: String) {
        val key = getOrGenKey()

        for ((index, char) in string.withIndex()) {
            val indexOfKey = charsMap[key[index % key.length]] ?: continue
            val indexOfChar = charsMap[char] ?: continue
            val decodedChar = reversedSquare[indexOfChar][indexOfKey]

            append(decodedChar)
        }
    }
}